contrib/storage-kafka/src/main/java/org/apache/drill/exec/store/kafka/KafkaRecordReader.java

```diff
@@ -0,0 +1,178 @@
+
+  private VectorContainerWriter writer;
+  private MessageReader messageReader;
+
+  private boolean unionEnabled;
+  private KafkaConsumer<byte[], byte[]> kafkaConsumer;
+  private KafkaStoragePlugin plugin;
+  private KafkaSubScanSpec subScanSpec;
+  private long kafkaPollTimeOut;
+  private long endOffset;
+
+  private long currentOffset;
+  private long totalFetchTime = 0;
+
+  private List<TopicPartition> partitions;
+  private final boolean enableAllTextMode;
+  private final boolean readNumbersAsDouble;
+
+  private Iterator<ConsumerRecord<byte[], byte[]>> messageIter;
+
+  public KafkaRecordReader(KafkaSubScan.KafkaSubScanSpec subScanSpec, List<SchemaPath> projectedColumns,
+      FragmentContext context, KafkaStoragePlugin plugin) {
+    setColumns(projectedColumns);
+    this.enableAllTextMode = context.getOptions().getOption(ExecConstants.KAFKA_ALL_TEXT_MODE).bool_val;
+    this.readNumbersAsDouble = context.getOptions()
+        .getOption(ExecConstants.KAFKA_READER_READ_NUMBERS_AS_DOUBLE).bool_val;
+    this.unionEnabled = context.getOptions().getOption(ExecConstants.ENABLE_UNION_TYPE);
+    this.plugin = plugin;
+    this.subScanSpec = subScanSpec;
+    this.endOffset = subScanSpec.getEndOffset();
+    this.kafkaPollTimeOut = Long.valueOf(plugin.getConfig().getDrillKafkaProps().getProperty(DRILL_KAFKA_POLL_TIMEOUT));
+  }
+
+  @Override
+  protected Collection<SchemaPath> transformColumns(Collection<SchemaPath> projectedColumns) {
+    Set<SchemaPath> transformed = Sets.newLinkedHashSet();
+    if (!isStarQuery()) {
+      for (SchemaPath column : projectedColumns) {
+        transformed.add(column);
+      }
+    } else {
+      transformed.add(Utilities.STAR_COLUMN);
+    }
+    return transformed;
+  }
+
+  @Override
+  public void setup(OperatorContext context, OutputMutator output) throws ExecutionSetupException {
+    this.writer = new VectorContainerWriter(output, unionEnabled);
+    messageReader = MessageReaderFactory.getMessageReader(plugin.getConfig().getDrillKafkaProps());
+    messageReader.init(context.getManagedBuffer(), Lists.newArrayList(getColumns()), this.writer,
+        this.enableAllTextMode, false, this.readNumbersAsDouble);
+    kafkaConsumer = messageReader.getConsumer(plugin);
+    partitions = Lists.newArrayListWithCapacity(1);
+    TopicPartition topicPartition = new TopicPartition(subScanSpec.getTopicName(), subScanSpec.getPartitionId());
+    partitions.add(topicPartition);
+    kafkaConsumer.assign(partitions);
+    currentOffset = subScanSpec.getStartOffset();
+    kafkaConsumer.seek(topicPartition, currentOffset);
+  }
+
+  private boolean hasNext() {
+    if (currentOffset > endOffset) {
+      logger.info("Read all the messages from Kafka. Start offset : {}, End offset : {}", currentOffset, endOffset);
+      return false;
+    } else if (messageIter != null && messageIter.hasNext()) {
+      return true;
+    } else {
+      return fetch();
+    }
```

Comment:
This mechanism seems a bit complex. Consider defining a higher-level message iterator class that hides the iterator/fetch details behind a uniform API. Let's call it a record iterator. That is:

* This reader simply calls `hasNext()`/`next()` on the record iterator.
* The record iterator loads the message iterator, iterates over it, and creates a new one for the next chunk.

One of the nice benefits of this approach is that you can unit test the record iterator without needing to run all of Drill.

Reply:
This has been refactored by creating MessageIterator class and moved this functionality to this class.